Mathlib/CategoryTheory/Triangulated/Opposite/Functor.lean

```diff
@@ -0,0 +1,195 @@
+that `F.mapTriangle.op` and `F.op.mapTriangle` correspond to each other via the
+equivalences `(Triangle C)ᵒᵖ ≌ Triangle Cᵒᵖ` and `(Triangle D)ᵒᵖ ≌ Triangle Dᵒᵖ`
+given by `CategoryTheory.Pretriangulated.triangleOpEquivalence`.
+
+-/
+
+namespace CategoryTheory
+
+variable {C D : Type*} [Category C] [Category D] [HasShift C ℤ] [HasShift D ℤ] (F : C ⥤ D)
+  [F.CommShift ℤ]
+
+namespace Pretriangulated.Opposite
+
+/-- If `F` commutes with shifts, so does `F.op`, for the shifts chosen on `Cᵒᵖ` in
+`CategoryTheory.Triangulated.Opposite.Basic`.
+-/
+noncomputable scoped instance commShiftOpInt : F.op.CommShift ℤ := by
+  letI F' : OppositeShift C ℤ ⥤ OppositeShift D ℤ := F.op
+  letI : F'.CommShift ℤ := F.commShiftOp ℤ
+  apply F'.commShiftPullback
+
+end Pretriangulated.Opposite
+
+namespace Functor
+
+open Category Limits Pretriangulated Opposite
+
+lemma op_commShiftIso_hom_app (X : Cᵒᵖ) (n m : ℤ) (h : n + m = 0):
+    (F.op.commShiftIso n).hom.app X =
+      (F.map ((shiftFunctorOpIso C n m h).hom.app X).unop).op ≫
+        ((F.commShiftIso m).inv.app X.unop).op ≫
+        (shiftFunctorOpIso D n m h).inv.app (op (F.obj X.unop)) := by
+  obtain rfl : m = -n := by omega
+  rfl
+
+lemma op_commShiftIso_inv_app (X : Cᵒᵖ) (n m : ℤ) (h : n + m = 0):
+    (F.op.commShiftIso n).inv.app X =
+      (shiftFunctorOpIso D n m h).hom.app (op (F.obj X.unop)) ≫
+        ((F.commShiftIso m).hom.app X.unop).op ≫
+          (F.map ((shiftFunctorOpIso C n m h).inv.app X).unop).op
+         := by
+  rw [← cancel_epi ((F.op.commShiftIso n).hom.app X), Iso.hom_inv_id_app,
+    op_commShiftIso_hom_app _ X n m h, assoc, assoc]
+  simp [← op_comp, ← F.map_comp]
+
+lemma shift_map_op {X Y : C} (f : X ⟶ Y) (n : ℤ) :
+    (F.map f).op⟦n⟧' = (F.op.commShiftIso n).inv.app _ ≫
+      (F.map (f.op⟦n⟧').unop).op ≫ (F.op.commShiftIso n).hom.app _ :=
+  (NatIso.naturality_1 (F.op.commShiftIso n) f.op).symm
+
+lemma map_shift_unop {X Y : Cᵒᵖ} (f : X ⟶ Y) (n : ℤ) :
+    F.map ((f⟦n⟧').unop) = ((F.op.commShiftIso n).inv.app Y).unop ≫
+      ((F.map f.unop).op⟦n⟧').unop ≫ ((F.op.commShiftIso n).hom.app X).unop := by
+  simp [shift_map_op]
+
+lemma map_opShiftFunctorEquivalence_unit_hom_app (F : C ⥤ D) [F.CommShift ℤ] (X : Cᵒᵖ) (n : ℤ) :
+    F.op.map ((opShiftFunctorEquivalence C n).unitIso.hom.app X) =
+      (opShiftFunctorEquivalence D n).unitIso.hom.app (F.op.obj X) ≫
+        (shiftFunctor D n).op.map (((F.op).commShiftIso n).inv.app X) ≫
+          ((F.commShiftIso n).hom.app _).op := by
+  rw [F.op_commShiftIso_inv_app _ _ _ (add_neg_cancel n)]
+  simp only [id_obj, op_obj, opShiftFunctorEquivalence, comp_obj, Iso.trans_hom, NatIso.op_hom,
+    isoWhiskerRight_hom, Iso.symm_hom, NatTrans.comp_app, NatTrans.op_app, whiskerRight_app, op_map,
+    unop_comp, Quiver.Hom.unop_op, map_comp, map_shiftFunctorCompIsoId_hom_app,
+    commShiftIso_hom_naturality_assoc, op_comp, assoc]
+  slice_rhs 2 3 => rw [← op_comp, ← Functor.map_comp, ← unop_comp, Iso.inv_hom_id_app]
+  simp
```

Comment:
This lemma is difficult to `rw` because the LHS (and RHS) gets simplified by `dsimp`, because of the `F.op.map`. Basically, `F.op.map` should almost never appear in a lemma because if we do `dsimp` (which should be mostly safe), we would have to apply a lemma in the backward direction. I suggest this version, and similarly for some other lemmas below:
```suggestion
@[reassoc]
lemma map_opShiftFunctorEquivalence_unitIso_hom_app_unop (X : Cᵒᵖ) (n : ℤ) :
    F.map ((opShiftFunctorEquivalence C n).unitIso.hom.app X).unop =
      (F.commShiftIso n).hom.app _ ≫
        (((F.op).commShiftIso n).inv.app X).unop⟦n⟧' ≫
        ((opShiftFunctorEquivalence D n).unitIso.hom.app (op _)).unop := by
  dsimp [opShiftFunctorEquivalence]
  simp only [map_comp, unop_comp, Quiver.Hom.unop_op, assoc,
    map_shiftFunctorCompIsoId_hom_app, commShiftIso_hom_naturality_assoc,
    op_commShiftIso_inv_app _ _ _ _ (add_neg_cancel n)]
  congr 3
  rw [← Functor.map_comp_assoc, ← unop_comp,
    Iso.inv_hom_id_app]
  dsimp
  rw [map_id, id_comp]
```